embulk-standards/src/main/java/org/embulk/standards/RenameFilterPlugin.java

```diff
@@ -82,12 +96,381 @@ public PageOutput open(TaskSource taskSource, Schema inputSchema,
+            throw new ConfigException("\"max_length\" in rule \"unique_number_suffix\" must be larger than " +(minimumMaxLengthInUniqueNumberSuffix-1));
+        }
+        if (max_length.isPresent() && digits.isPresent() && max_length.get() < digits.get()) {
+            throw new ConfigException("\"max_length\" in rule \"unique_number_suffix\" must be larger than \"digits\"");
+        }
+        int digitsOfNumberOfColumns = Integer.toString(inputSchema.getColumnCount() + offset - 1).length();
+        if (max_length.isPresent() && max_length.get() <= digitsOfNumberOfColumns) {
+            throw new ConfigException("\"max_length\" in rule \"unique_number_suffix\" must be larger than digits of ((number of columns) + \"offset\" - 1)");
+        }
+        if (digits.isPresent() && digits.get() <= digitsOfNumberOfColumns) {
+            throw new ConfigException("\"digits\" in rule \"unique_number_suffix\" must be larger than digits of ((number of columns) + \"offset\" - 1)");
+        }
+
+        // Columns should not be truncated here initially. Uniqueness should be identified before truncated.
+
+        // Iterate for initial states.
+        HashSet<String> originalColumnNames = new HashSet<>();
+        HashMap<String, Integer> columnNameCountups = new HashMap<>();
+        for (Column column : inputSchema.getColumns()) {
+            originalColumnNames.add(column.getName());
+            columnNameCountups.put(column.getName(), offset);
+        }
+
+        Schema.Builder outputBuilder = Schema.builder();
+
+        HashSet<String> fixedColumnNames = new HashSet<>();
+        for (Column column : inputSchema.getColumns()) {
+            String truncatedName = column.getName();
+            if (column.getName().length() > max_length.or(Integer.MAX_VALUE)) {
+                truncatedName = column.getName().substring(0, max_length.get());
+            }
+
+            // Fix with the new name candidate if the new name does not conflict with the fixed names on the left.
+            // Conflicts with original names do not matter here.
+            if (!fixedColumnNames.contains(truncatedName)) {
+                // The original name is counted up.
+                columnNameCountups.put(column.getName(), columnNameCountups.get(column.getName()) + 1);
+                // The truncated name is fixed.
+                fixedColumnNames.add(truncatedName);
+                outputBuilder.add(truncatedName, column.getType());
+                continue;
+            }
+
+            int index = columnNameCountups.get(column.getName());
+            if (index <= offset) {
+                // It may happen if |esteem_original_names| is false.
+                // See |TestRenameFilterPlugin#checkUniqueNumberSuffixRule7WithoutEsteemOriginalNames|.
+                index = offset + 1;
+            }
+            String concatenatedName;
+            do {
+                // String#format is not used to avoid locale-related problems.
```

Comment:
Thanks! Good to know.

But finally, because Java's format doesn't have `"%*d"` (variable width), `String#format` did not make it very simple. :(  I kept this part unchanged.

```
final String differentiatorFormat = String.format(
    Locale.ENGLISH,
    "%%s%%%sd",
    digits.isPresent() ? ("0" + Integer.toString(digits.get())) : "");
final String differentiatorString = String.format(
    Locale.ENGLISH, differentiatorFormat, delimiter, index);
```